refactor: common vitest configurations (#1107)
code-pushup/cli: tools/README.md

## Vitest config factory and setup presets

> This folder contains utilities to centralize and standardize Vitest configuration across the monorepo.

Reviewer: ```suggestion
This project contains utilities to centralize and standardize Vitest configuration across a monorepo.
```

### Files

- `vitest-config-factory.ts`: builds typed Vitest configs with sensible defaults.
- `vitest-setup-presets.ts`: reusable groups of `setupFiles` paths by test kind.

### Goals

- Reduce duplication across `vitest.*.config.ts` files.
- Keep per-package intent clear with minimal overrides.

> - Provide safe defaults and easy extension points.

Reviewer: We could remove it as its already mentioned in first paragraph

### Quick start

Use the factory in your suite configs (project root is required):

```ts
/// <reference types="vitest" />
import { createE2eConfig, createIntConfig, createUnitConfig } from '../../tools/vitest-config-factory.js';

export default createUnitConfig('core', {
  projectRoot: new URL('../../', import.meta.url),
});
```

Creators:

- `createUnitConfig(projectKey, { projectRoot, ...options })`
- `createIntConfig(projectKey, { projectRoot, ...options })`
- `createE2eConfig(projectKey, { projectRoot, ...options })`

> `projectKey` is used for cache and coverage directories.

Reviewer: You could give an example

### Defaults

Common to all kinds:

- `reporters: ['basic']`, `globals: true`, `environment: 'node'`
- `alias: tsconfigPathAliases()`
- `pool: 'threads'` with `singleThread: true`
- Cache directories resolved from `projectRoot` (absolute paths)

Coverage:

- Unit/Int: enabled by default, reports to `<projectRoot>/coverage/<project>/<kind>-tests`
- E2E: disabled by default
- Default exclude: `['mocks/**', '**/types.ts']`

Global setup:

- Unit/Int: `['<projectRoot>/global-setup.ts']` by default
- E2E: none by default (set per-suite if needed)

Include patterns:

- Unit: `src/**/*.unit.test.*`
- Int: `src/**/*.int.test.*`

> - E2E: `tests/**/*.e2e.test.*`

Reviewer: This information is tightly coupled to our implementation. MAybe link the perset directly here instead of repeating?

### setupFiles strategy

Baseline `setupFiles` are injected automatically by kind:

- Unit baseline: `console.mock.ts`, `reset.mocks.ts`
- Int baseline: `console.mock.ts`, `reset.mocks.ts`
- E2E baseline: `reset.mocks.ts`

Extend with additional files using `options.setupFiles` — they append after the baseline (paths are project-root-relative):

```ts
export default createUnitConfig('core', {
  projectRoot: new URL('../../', import.meta.url),
  setupFiles: ['testing/test-setup/src/lib/cliui.mock.ts'],
});

> ```

Reviewer: Same here, strongly coupled to implementation that can change often.

```ts
import { setupPresets } from '../../tools/vitest-setup-presets.js';

export default createIntConfig('core', {
  projectRoot: new URL('../../', import.meta.url),
  setupFiles: [...setupPresets.int.portalClient],
});
```

Preset keys:

- `setupPresets.unit.{base,cliui,fs,git,portalClient,matchersCore,matcherPath}`
- `setupPresets.int.{base,cliui,fs,git,portalClient,matcherPath,chromePath}`
- `setupPresets.e2e.{base}`

### Options reference

`CreateVitestConfigOptions` (required + optional):

- `projectKey` (string): coverage/cache naming.
- `kind` ('unit' | 'int' | 'e2e'): test kind.
- `projectRoot` (string | URL): absolute root for all paths.
- `include?: string[]`: override default include globs.
- `setupFiles?: string[]`: extra setup files (appended to baseline; project-root-relative).
- `overrideSetupFiles?: boolean`: skip baseline and use only provided list.
- `globalSetup?: string[]`: override default global setup (project-root-relative).
- `coverage?: { enabled?, exclude?, reportsSubdir? }`
- `testTimeout?: number`: e.g., for E2E.
- `typecheckInclude?: string[]`: include patterns for Vitest typecheck.

> - `cacheKey?: string`: custom cache dir suffix.

Reviewer: The best DX is given if the user does no need to lear another type. 

I would suggest try to keep the types as close as possible to existing virest types. 

e.g. :  `createIntConfig({/*vitestTestConfig*/}, { projectKey: 'unit', /*other custom options we need that are not already inside vitest types */})`
